[CL-894] Create HomePage model

nTraum opened the pull request:
This commit introduces the table structure for the home page model. Most of it comes from the existing settings in the `AppConfiguration`, including the database defaults. The more thorough validations will come in CL-897.

## Checklist

- [ ] Added entry to changelog
<details>
<summary>More info</summary>
Add a concise line to the 'Next release' section of the changelog (docs/README.md) so people other than developers can understand what has changed where. E.g. 'Added an error message to the project name field of the project edit form (Admin > Projects > Edit)'.
</details>

- [ ] WCAG 2.1 AA proof
<details>
<summary>More info</summary>
For front-end devs only. Is your work conforming with the WCAG 2.1 AA rules? If you need more info, read the [a11y page](https://www.notion.so/citizenlab/a11y-7568f83d42ab4895ac133b89d358997b) on our Notion.
</details>

- [ ] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [ ] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

## Links

- [citizenlab-ee PR](**put URL here** or remove)
- [Specs](**put URL here** or remove)
- [Epic Deployment](**put URL here** or remove)

## How urgent is a code review?

Let the reviewer(s) know how urgent the code review is, so they can prioritize their work accordingly. Be specific (e.g. by Wednesday, end of the day/this week/... is better than 'urgent' or 'very urgent'). Optionally provide a word of explanation on your deadline.

nTraum committed:
[CL-894] Create HomePage model

This commit introduces the table structure for the home page model. Most of it comes from the existing settings in the `AppConfiguration`, including the database defaults. The more thorough validations will come in CL-897.

alexander-cit commented on back/db/migrate/20220531123916_create_home_pages.rb:
```diff
@@ -0,0 +1,30 @@
+    create_table :home_pages, id: :uuid do |t|
+      t.boolean :header_avatars_enabled, default: true, null: false
+
+      t.boolean :top_info_section_enabled, default: false, null: false
```
Don't we need to add `hero_banner_enabled` and `projects_enabled` (probably) too?

nTraum replied:
Yes, will add!

alexander-cit commented on back/db/migrate/20220531123916_create_home_pages.rb:
```diff
@@ -0,0 +1,30 @@
+class CreateHomePages < ActiveRecord::Migration[6.1]
+  def change
+    create_table :home_pages, id: :uuid do |t|
+      t.boolean :header_avatars_enabled, default: true, null: false
```
Is it fka `display_header_avatars`? On the design, we still use `Display avatars`?

nTraum replied:
What does `fka` mean? Do you suggest changing the name of the column?

alexander-cit replied:
Formerly known as :) 

> Do you suggest changing the name of the column?

Not really suggesting, just wondering why you renamed `display_header_avatars` to `header_avatars_enabled`.
Perfectly fine as it is now :)

nTraum replied:
I wanted to make the toggles consistent to end with `enabled`.

alexander-cit replied:
Maybe I would try to distinguish section toggles and other toggles, but again, it's fine as it is now :smile:

alexander-cit commented on back/db/migrate/20220531123916_create_home_pages.rb:
```diff
@@ -0,0 +1,30 @@
+
+class CreateHomePages < ActiveRecord::Migration[6.1]
+  def change
+    create_table :home_pages, id: :uuid do |t|
```
Will we add fields like `signedOutHeaderOverlayColor` from `back/config/schemas/style.schema.json`? They are on the design too.

nTraum replied:
Ah, I didn't know about this one! Yes will add.

nTraum replied:
I'm also missing the image itself, because I think this is not the AppConfig currently.

nTraum committed:
Add projects_enabled and banner_enabled toggles for HomePage model

nTraum committed:
Default to enabled info and bottom sections for HomePage model

nTraum committed:
Add image color and opacity as well

nTraum committed:
Add projects_header multiloc to HomePage

jinjagit commented on back/db/migrate/20220531123916_create_home_pages.rb:
```diff
@@ -0,0 +1,36 @@
+      t.boolean :banner_enabled, default: true, null: false
+      t.string :banner_layout, default: 'full_width_banner_layout', null: false
+      t.string :banner_signed_in_header
+      t.string :banner_signed_in_text
```
Is this not a multiloc? If not, shouldn't it be? (And same for any other text that can be set for customisable content)

alexander-cit replied:
If it's the former `customized_button#text`, then yes, it should be multiloc.
And to avoid confusion in the future (working on this epic), it can be useful to describe the mapping of old and new fields. Or even better, preserve old names.

nTraum replied:
I'll make this a multiloc, good catch Simon!

I think the naming can be improved, which is why I took the liberty to do so here: For example I think it makes sense to prefix all banner related settings with `banner`, because it brings clarity in what the setting belongs to.

I'll add the new and old tables to Notion in a second, good idea Alex!

jinjagit reviewed (approved):
LGTM - excluding the comments from Alex. I just have one question about using simple strings for customisable content and not multilocs.

nTraum committed:
Rename header_avatars_enabled to banner_avatars_enabled, because it belongs to the banner.

Make banner_signed_in_header, banner_signed_in_text, banner_signed_out_header, banner_signed_out_subheader and banner_signed_out_text jsonb fields, they are multiloc.
